src/internal_modules/roc_core/mpsc_queue_impl.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_core/mpsc_queue_impl.h"
+#include "roc_core/atomic_ops.h"
+#include "roc_core/cpu_instructions.h"
+
+namespace roc {
+namespace core {
+
+MpscQueueImpl::MpscQueueImpl()
+    : tail_(&stub_)
+    , head_(&stub_) {
+}
+
+MpscQueueImpl::~MpscQueueImpl() {
+    if (head_ != &stub_) {
+        roc_panic("mpsc_queue: queue isn't empty on destruct");
```

Comment:
```suggestion
        roc_panic("mpsc queue: queue isn't empty on destruct");
```

(we usually use format like this in logs and panics)

Reply:
My bad. :face_palm